### PR TITLE
Send full system data on registration

### PR DESCRIPTION
**Pull request by Itxaka**
This uses the ghw library to send the full system data during the registration.

This sends, Net, Block, CPU, memory, chassis, etc... data in a nicely structured block which makes it easy to use for the inventory.

This could potentially substitute the smbios data as it overlaps nicely with the same data that smbios provides, and would help us in not having to fully parse the smbios data ourselves like we do.

Signed-off-by: Itxaka <igarcia@suse.com>

### CODE REVIEW

**Comment by Itxaka** [comment]
This is an example of the data provided from a qemu instance:

```
&hostinfo.HostInfo{
  Memory: &memory.Info{
    Area: memory.Area{
      TotalPhysicalBytes: 12616466432,
      TotalUsableBytes: 12226326528,
      SupportedPageSizes: []uint64{
        1073741824,
        2097152,
      },
      Modules: nil,
    },
  },
  Block: &block.Info{
    TotalPhysicalBytes: 43623579648,
    Disks: []*block.Disk{
      &block.Disk{
        Name: "sr0",
        SizeBytes: 673906688,
        PhysicalBlockSizeBytes: 2048,
        DriveType: 3,
        IsRemovable: true,
        StorageController: 2,
        BusPath: "pci-0000:00:1f.2-ata-1.0",
        NUMANodeID: 0,
        Vendor: "QEMU",
        Model: "QEMU_DVD-ROM",
        SerialNumber: "QM00001",
        WWN: "unknown",
        Partitions: []*block.Partition{},
      },
      &block.Disk{ // p0
        Name: "vda",
        SizeBytes: 42949672960,
        PhysicalBlockSizeBytes: 512,
        DriveType: 1,
        IsRemovable: false,
        StorageController: 4,
        BusPath: "pci-0000:04:00.0",
        NUMANodeID: 0,
        Vendor: "0x1af4",
        Model: "unknown",
        SerialNumber: "unknown",
        WWN: "unknown",
        Partitions: []*block.Partition{
          &block.Partition{
            Disk: p0,
            Name: "vda1",
            Label: "COS_GRUB",
            MountPoint: "",
            SizeBytes: 67108864,
            Type: "vfat",
            IsReadOnly: true,
            UUID: "16c4870a-52b8-4802-856b-5adadefef89f",
          },
          &block.Partition{
            Disk: p0,
            Name: "vda2",
            Label: "COS_OEM",
            MountPoint: "",
            SizeBytes: 67108864,
            Type: "ext4",
            IsReadOnly: true,
            UUID: "7bac8569-74cd-4b5a-a963-cd14af3fbaed",
          },
          &block.Partition{
            Disk: p0,
            Name: "vda3",
            Label: "COS_RECOVERY",
            MountPoint: "",
            SizeBytes: 8589934592,
            Type: "ext4",
            IsReadOnly: true,
            UUID: "81b08ac1-2893-406e-aa54-d0dce8e0711d",
          },
          &block.Partition{
            Disk: p0,
            Name: "vda4",
            Label: "COS_STATE",
            MountPoint: "",
            SizeBytes: 16106127360,
            Type: "ext4",
            IsReadOnly: true,
            UUID: "0ee3e931-098c-40ab-a9cf-1b12594745fb",
          },
          &block.Partition{
            Disk: p0,
            Name: "vda5",
            Label: "COS_PERSISTENT",
            MountPoint: "",
            SizeBytes: 18117296128,
            Type: "ext4",
            IsReadOnly: true,
            UUID: "adcf76a9-dce8-42b5-9a4f-f46751c86bb7",
          },
        },
      },
    },
    Partitions: nil,
  },
  CPU: &cpu.Info{
    TotalCores: 2,
    TotalThreads: 2,
    Processors: []*cpu.Processor{
      &cpu.Processor{
        ID: 0,
        NumCores: 1,
        NumThreads: 1,
        Vendor: "GenuineIntel",
        Model: "Intel(R) Core(TM) i7-7700K CPU @ 4.20GHz",
        Capabilities: []string{
          "fpu",
          "vme",
          "de",
          "pse",
          "tsc",
          "msr",
          "pae",
          "mce",
          "cx8",
          "apic",
          "sep",
          "mtrr",
          "pge",
          "mca",
          "cmov",
          "pat",
          "pse36",
          "clflush",
          "mmx",
          "fxsr",
          "sse",
          "sse2",
          "ss",
          "syscall",
          "nx",
          "pdpe1gb",
          "rdtscp",
          "lm",
          "constant_tsc",
          "arch_perfmon",
          "rep_good",
          "nopl",
          "xtopology",
          "cpuid",
          "tsc_known_freq",
          "pni",
          "pclmulqdq",
          "vmx",
          "ssse3",
          "fma",
          "cx16",
          "pdcm",
          "pcid",
          "sse4_1",
          "sse4_2",
          "x2apic",
          "movbe",
          "popcnt",
          "tsc_deadline_timer",
          "aes",
          "xsave",
          "avx",
          "f16c",
          "rdrand",
          "hypervisor",
          "lahf_lm",
          "abm",
          "3dnowprefetch",
          "cpuid_fault",
          "invpcid_single",
          "pti",
          "ssbd",
          "ibrs",
          "ibpb",
          "stibp",
          "tpr_shadow",
          "vnmi",
          "flexpriority",
          "ept",
          "vpid",
          "ept_ad",
          "fsgsbase",
          "tsc_adjust",
          "bmi1",
          "avx2",
          "smep",
          "bmi2",
          "erms",
          "invpcid",
          "mpx",
          "rdseed",
          "adx",
          "smap",
          "clflushopt",
          "xsaveopt",
          "xsavec",
          "xgetbv1",
          "xsaves",
          "arat",
          "umip",
          "md_clear",
          "arch_capabilities",
        },
        Cores: []*cpu.ProcessorCore{
          &cpu.ProcessorCore{
            ID: 0,
            Index: 0,
            NumThreads: 1,
            LogicalProcessors: []int{
              0,
            },
          },
        },
      },
      &cpu.Processor{
        ID: 1,
        NumCores: 1,
        NumThreads: 1,
        Vendor: "GenuineIntel",
        Model: "Intel(R) Core(TM) i7-7700K CPU @ 4.20GHz",
        Capabilities: []string{
          "fpu",
          "vme",
          "de",
          "pse",
          "tsc",
          "msr",
          "pae",
          "mce",
          "cx8",
          "apic",
          "sep",
          "mtrr",
          "pge",
          "mca",
          "cmov",
          "pat",
          "pse36",
          "clflush",
          "mmx",
          "fxsr",
          "sse",
          "sse2",
          "ss",
          "syscall",
          "nx",
          "pdpe1gb",
          "rdtscp",
          "lm",
          "constant_tsc",
          "arch_perfmon",
          "rep_good",
          "nopl",
          "xtopology",
          "cpuid",
          "tsc_known_freq",
          "pni",
          "pclmulqdq",
          "vmx",
          "ssse3",
          "fma",
          "cx16",
          "pdcm",
          "pcid",
          "sse4_1",
          "sse4_2",
          "x2apic",
          "movbe",
          "popcnt",
          "tsc_deadline_timer",
          "aes",
          "xsave",
          "avx",
          "f16c",
          "rdrand",
          "hypervisor",
          "lahf_lm",
          "abm",
          "3dnowprefetch",
          "cpuid_fault",
          "invpcid_single",
          "pti",
          "ssbd",
          "ibrs",
          "ibpb",
          "stibp",
          "tpr_shadow",
          "vnmi",
          "flexpriority",
          "ept",
          "vpid",
          "ept_ad",
          "fsgsbase",
          "tsc_adjust",
          "bmi1",
          "avx2",
          "smep",
          "bmi2",
          "erms",
          "invpcid",
          "mpx",
          "rdseed",
          "adx",
          "smap",
          "clflushopt",
          "xsaveopt",
          "xsavec",
          "xgetbv1",
          "xsaves",
          "arat",
          "umip",
          "md_clear",
          "arch_capabilities",
        },
        Cores: []*cpu.ProcessorCore{
          &cpu.ProcessorCore{
            ID: 0,
            Index: 0,
            NumThreads: 1,
            LogicalProcessors: []int{
              1,
            },
          },
        },
      },
    },
  },
  Topology: &topology.Info{
    Architecture: 0,
    Nodes: []*topology.Node{
      &topology.Node{
        ID: 0,
        Cores: []*cpu.ProcessorCore{
          &cpu.ProcessorCore{
            ID: 0,
            Index: 0,
            NumThreads: 2,
            LogicalProcessors: []int{
              0,
              1,
            },
          },
        },
        Caches: []*memory.Cache{
          &memory.Cache{
            Level: 1,
            Type: 1,
            SizeBytes: 32768,
            LogicalProcessors: []uint32{
              0,
            },
          },
          &memory.Cache{
            Level: 1,
            Type: 1,
            SizeBytes: 32768,
            LogicalProcessors: []uint32{
              1,
            },
          },
          &memory.Cache{
            Level: 1,
            Type: 2,
            SizeBytes: 32768,
            LogicalProcessors: []uint32{
              0,
            },
          },
          &memory.Cache{
            Level: 1,
            Type: 2,
            SizeBytes: 32768,
            LogicalProcessors: []uint32{
              1,
            },
          },
          &memory.Cache{
            Level: 2,
            Type: 0,
            SizeBytes: 4194304,
            LogicalProcessors: []uint32{
              0,
            },
          },
          &memory.Cache{
            Level: 2,
            Type: 0,
            SizeBytes: 4194304,
            LogicalProcessors: []uint32{
              1,
            },
          },
          &memory.Cache{
            Level: 3,
            Type: 0,
            SizeBytes: 16777216,
            LogicalProcessors: []uint32{
              0,
            },
          },
          &memory.Cache{
            Level: 3,
            Type: 0,
            SizeBytes: 16777216,
            LogicalProcessors: []uint32{
              1,
            },
          },
        },
        Distances: []int{
          10,
        },
        Memory: &memory.Area{
          TotalPhysicalBytes: 12616466432,
          TotalUsableBytes: 12226322432,
          SupportedPageSizes: []uint64{
            1073741824,
            2097152,
          },
          Modules: nil,
        },
      },
    },
  },
  Network: &net.Info{
    NICs: []*net.NIC{
      &net.NIC{
        Name: "eth0",
        MacAddress: "52:54:00:e7:48:ac",
        IsVirtual: false,
        Capabilities: []*net.NICCapability{
          &net.NICCapability{
            Name: "rx-checksumming",
            IsEnabled: true,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "tx-checksumming",
            IsEnabled: true,
            CanEnable: true,
          },
          &net.NICCapability{
            Name: "tx-checksum-ipv4",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "tx-checksum-ip-generic",
            IsEnabled: true,
            CanEnable: true,
          },
          &net.NICCapability{
            Name: "tx-checksum-ipv6",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "tx-checksum-fcoe-crc",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "tx-checksum-sctp",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "scatter-gather",
            IsEnabled: true,
            CanEnable: true,
          },
          &net.NICCapability{
            Name: "tx-scatter-gather",
            IsEnabled: true,
            CanEnable: true,
          },
          &net.NICCapability{
            Name: "tx-scatter-gather-fraglist",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "tcp-segmentation-offload",
            IsEnabled: true,
            CanEnable: true,
          },
          &net.NICCapability{
            Name: "tx-tcp-segmentation",
            IsEnabled: true,
            CanEnable: true,
          },
          &net.NICCapability{
            Name: "tx-tcp-ecn-segmentation",
            IsEnabled: true,
            CanEnable: true,
          },
          &net.NICCapability{
            Name: "tx-tcp-mangleid-segmentation",
            IsEnabled: false,
            CanEnable: true,
          },
          &net.NICCapability{
            Name: "tx-tcp6-segmentation",
            IsEnabled: true,
            CanEnable: true,
          },
          &net.NICCapability{
            Name: "generic-segmentation-offload",
            IsEnabled: true,
            CanEnable: true,
          },
          &net.NICCapability{
            Name: "generic-receive-offload",
            IsEnabled: true,
            CanEnable: true,
          },
          &net.NICCapability{
            Name: "large-receive-offload",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "rx-vlan-offload",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "tx-vlan-offload",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "ntuple-filters",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "receive-hashing",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "highdma",
            IsEnabled: true,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "rx-vlan-filter",
            IsEnabled: true,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "vlan-challenged",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "tx-lockless",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "netns-local",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "tx-gso-robust",
            IsEnabled: true,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "tx-fcoe-segmentation",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "tx-gre-segmentation",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "tx-gre-csum-segmentation",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "tx-ipxip4-segmentation",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "tx-ipxip6-segmentation",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "tx-udp_tnl-segmentation",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "tx-udp_tnl-csum-segmentation",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "tx-gso-partial",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "tx-tunnel-remcsum-segmentation",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "tx-sctp-segmentation",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "tx-esp-segmentation",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "tx-udp-segmentation",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "tx-gso-list",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "fcoe-mtu",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "tx-nocache-copy",
            IsEnabled: false,
            CanEnable: true,
          },
          &net.NICCapability{
            Name: "loopback",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "rx-fcs",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "rx-all",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "tx-vlan-stag-hw-insert",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "rx-vlan-stag-hw-parse",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "rx-vlan-stag-filter",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "l2-fwd-offload",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "hw-tc-offload",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "esp-hw-offload",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "esp-tx-csum-hw-offload",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "rx-udp_tunnel-port-offload",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "tls-hw-tx-offload",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "tls-hw-rx-offload",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "rx-gro-hw",
            IsEnabled: true,
            CanEnable: true,
          },
          &net.NICCapability{
            Name: "tls-hw-record",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "rx-gro-list",
            IsEnabled: false,
            CanEnable: true,
          },
          &net.NICCapability{
            Name: "macsec-hw-offload",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "rx-udp-gro-forwarding",
            IsEnabled: false,
            CanEnable: true,
          },
          &net.NICCapability{
            Name: "hsr-tag-ins-offload",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "hsr-tag-rm-offload",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "hsr-fwd-offload",
            IsEnabled: false,
            CanEnable: false,
          },
          &net.NICCapability{
            Name: "hsr-dup-offload",
            IsEnabled: false,
            CanEnable: false,
          },
        },
        PCIAddress: nil,
      },
    },
  },
  GPU: &gpu.Info{
    GraphicsCards: []*gpu.GraphicsCard{
      &gpu.GraphicsCard{
        Address: "virtio0",
        Index: 0,
        DeviceInfo: nil,
        Node: nil,
      },
    },
  },
  Chassis: &chassis.Info{
    AssetTag: "",
    SerialNumber: "",
    Type: "1",
    TypeDescription: "Other",
    Vendor: "QEMU",
    Version: "pc-q35-7.1",
  },
  BIOS: &bios.Info{
    Vendor: "EFI Development Kit II / OVMF",
    Version: "0.0.0",
    Date: "02/06/2015",
  },
  Baseboard: &baseboard.Info{
    AssetTag: "unknown",
    SerialNumber: "unknown",
    Vendor: "unknown",
    Version: "unknown",
    Product: "unknown",
  },
  Product: &product.Info{
    Family: "",
    Name: "Standard PC (Q35 + ICH9, 2009)",
    Vendor: "QEMU",
    SerialNumber: "",
    UUID: "79cfdb39-bc36-41d6-a2b2-5726b6bca234",
    SKU: "",
    Version: "pc-q35-7.1",
  },
} 

```

**Comment by Itxaka** [comment]
Of special interest IMO is the GPU stuff (You can identify the nodes with GPUs for GPU-related taks) and the NIC capabilities.

**Comment by Itxaka** [comment]
This doesnt use the data in any way for now, its only a POC to see if we would like this approach and how we can take advantage of the extra data.

Also to notice if there is something missing that we can contribute back to ghw that would help us here, for example I can see that memory is not reporting the modules...

**Comment by kkaempf** [comment]
Visualizing this in the UI will be the hard part ...

**Comment by Itxaka** [comment]
> Visualizing this in the UI will be the hard part ...

I been wondering how to store this somehow.... Maybe a secret associated with the inventory could be created and the json data stored in there, so its available for the selector AND maybe the labels if we want. So its checked JIT when the selector kicks in?

There is always the possibility of adding a custom resource to save this data....and just have it as another item in the inventory controller, so no need to show it in the UI directly under inventories, it could have its own resource like Plans. But that would be too much IMO, the less resources created the better.

Going with the secret linked to this would be much nicer IMO, we could even start by just exposing this data to the label templating so you can select what data you want to store on nodes (i.e. not sure you need ALL this data, but maybe you want to store only the GPU name or the Network capabilities for example)

Still think its vary valuable data to expose, but not sure of its use, where to store it and so on. Something to discuss in the planning IMO, so we have a clear path for the future of the node data, but its clear we need to do something with it, as the use-cases are clear for more hardware data AND for more selector flexibility.

**Comment by kkaempf** [comment]
`ros-operator` put the whole (well, `dmicode`) tree into the `machineInventory` CRD ...

**Comment by Itxaka** [comment]
> `ros-operator` put the whole (well, `dmicode`) tree into the `machineInventory` CRD ...

yes, and it was terrible and if you described the machineinventory from cli you will get a 2Mb yaml in your cli lol

Its about a better and nicer implementation of the same basic idea :)

File: pkg/register/websocket.go [hunk 1]
```diff
@@ -32,6 +32,7 @@ const (
 	MsgUndefined MessageType = iota
 	MsgReady
 	MsgSmbios
+	MsgSystemData
 	MsgLabels
 	MsgGet
 	MsgLast = MsgGet // MsgLast must point to the last message
```
Comment (fgiudici) [inline pkg/register/websocket.go:38]: The new message should be put just before MsgLast (and MsgLast should be updated to MsgSystemData).
This way we preserve old messages.
Something like:

```
...
MsgSmbios
MsgLabels
MsgGet
MsgSystemData
MsgLast = MsgSystemData // MsgLast must point to the last message

```

File: pkg/register/register.go [hunk 2]
```diff
@@ -126,3 +133,19 @@ func sendData(conn *websocket.Conn, smbios bool) error {
...
+	if smbios {
+		logrus.Debug("send system data")
+		data, err := hostinfo.Host(ghw.WithDisableWarnings())
+		if err != nil {
+			return errors.Wrap(err, "failed to read system data")
+		}
+		logrus.Debugf("seding the following data:\n%s", litter.Sdump(data))
+		err = SendJSONData(conn, MsgSystemData, data)
+		if err != nil {
+			return fmt.Errorf("sending SMBIOS data: %w", err)
```
Comment (fgiudici) [inline pkg/register/register.go:147]:
```suggestion
			return fmt.Errorf("sending system data: %w", err)
```

File: pkg/register/websocket.go (hunk 1 above)
Comment (Itxaka) [reply pkg/register/websocket.go:38]: thanks!

**Review by fgiudici — COMMENTED** [review]
Love this ❤️ 
Yep, we have to discuss this, this could be really useful!

**Comment by fgiudici** [comment]
> I been wondering how to store this somehow.... Maybe a secret associated with the inventory could be created and the json data stored in there, so its available for the selector AND maybe the labels if we want. So its checked JIT when the selector kicks in?

I would consider a ConfigMap that we bind to the MachineInventory !

**Comment by agracey** [comment]
We should likely keep a way to optionally turn off the storage of the data. I could see it quickly becoming a scaling bottleneck for larger amounts of nodes due to etcd.

**Comment by agracey** [comment]
I know it's not a popular opinion but maybe an RDBMS would be more appropriate here than putting it all in the k8s api?

File: pkg/register/register.go [hunk 3]
```diff
@@ -126,3 +133,19 @@ func sendData(conn *websocket.Conn, smbios bool) error {
...
+
+func sendSystemData(conn *websocket.Conn, smbios bool) error {
+	if smbios {
+		logrus.Debug("send system data")
+		data, err := hostinfo.Host(ghw.WithDisableWarnings())
+		if err != nil {
+			return errors.Wrap(err, "failed to read system data")
+		}
+		logrus.Debugf("seding the following data:\n%s", litter.Sdump(data))
+		err = SendJSONData(conn, MsgSystemData, data)
```
Comment (alexander-demicev) [inline pkg/register/register.go:145]:
```suggestion
		if err := SendJSONData(conn, MsgSystemData, data); err != nil {
```

**Review by alexander-demicev — COMMENTED** [review]
that looks really useful, let's figure out how to store data

**Comment by Itxaka** [comment]
> I know it's not a popular opinion but maybe an RDBMS would be more appropriate here than putting it all in the k8s api?

That would make us require an extra requirements to deploy. And not a small one, having a db around with all its issues (backup, managing it, permissions, etc...).

Yikes.

We would only need to store this if we expect the selector to be more flexible, i.e. allow selecting with hardware data. Plus we dont really need to store this if we dont have any selector plan. Only a POC to see what kind of data we can get from this, and maybe use it to create extra labels on the node with this data (Network, disks, memory, gpu, etc...?)

I think that the nodes we register should have as much data as possible to identify them, it would be nice to have that, so this is just a first step and a POC so we can discuss this. 

I am agaisnt saving all this data anywhere, as it creates more issues that it resolves. HAving it a secret or a configmap its good until you have hundreds of them :/

IMHO the idea here should be to gather as much data as possible on the registration part, as that its basically free, and then decide what we wanna do with it. Labels its the way to go IMHO, but we could always also expand the inventory struct to contains a Hardware struct that includes the most valued data. But still....labels allows for a lot of flexibility across the sholw k8s standard. Search, selectors, etc.. are all compatible with labels out of the box...

**Comment by agracey** [comment]
> That would make us require an extra requirements to deploy. And not a small one, having a db around with all its issues (backup, managing it, permissions, etc...).

Fair. But we just need to recognize the tradeoffs of just shoving anything and everything into the kubeapi/etcd.

> IMHO the idea here should be to gather as much data as possible on the registration part, as that its basically free, and then decide what we wanna do with it 

That makes sense to me. I wonder if [shoving registration data into an event ](https://book-v1.book.kubebuilder.io/beyond_basics/creating_events.html)would be interesting. That would allow the admin to consume the data while not having to persistently store it.

**Comment by Itxaka** [comment]
```
$ kubectl describe machineinventory --all-namespaces
Name:         my-machine
Namespace:    fleet-default
Labels:       BlockDevice0-ControllerType=SCSI
              BlockDevice0-DriveType=ODD
              BlockDevice0-Name=sr0
              BlockDevice0-Removable=true
              BlockDevice0-Size=754712576
              BlockDevice1-ControllerType=virtio
              BlockDevice1-DriveType=HDD
              BlockDevice1-Name=vda
              BlockDevice1-Removable=false
              BlockDevice1-Size=32212254720
              BlockTotal=2
              CpuModel=Intel-R--Core-TM--i7-7700K-CPU---4-20GHz
              CpuTotalCores=7
              CpuTotalThreads=7
              CpuVendor=GenuineIntel
              NetIface0-Name=eth0
              NetIface0-Virtual=false
              NetNumberInterfaces=1
              TotalMemory=12616466432
              location=europe
              machineUUID=40227f83-f8c4-49b8-9899-16ba2878498a
              manufacturer=QEMU
              productName=Standard-PC-Q35-ICH9-2009
              serialNumber=Not-Specified
Annotations:  <none>
API Version:  elemental.cattle.io/v1beta1
```

**Comment by Itxaka** [comment]
on latest patch I added the prefix `elemental-cattle.io/` to the labels to make sure they are tracked to us.

**Review by davidcassany — COMMENTED** [review]
Looks good to me.

I am not sure we need to keep all the raw system data anywhere and this is something that could be revisited in the future if needed. To my eyes the important thing is that the data is sent and processed to include some extra labels into the inventory. I am fine with the set of labels you suggested, some generic data to identify hw. We could easily extend it if we see the need for some extra labels.

**Comment by Itxaka** [comment]
@fgiudici comments?

**Comment by fgiudici** [comment]
> @fgiudici comments?

just one: wonderful!!

**Review by fgiudici — APPROVED** [review]

**Pull request merged**